lambda-runtime/src/AWS/Lambda/Client.hs

```diff
@@ -0,0 +1,175 @@
+import Control.Monad.Except (ExceptT(..), liftEither, throwError, withExceptT)
+import System.Environment (lookupEnv)
+
+import qualified Data.Aeson          as JSON
+import qualified Data.ByteString     as BS
+import qualified Network.HTTP.Client as HTTP
+import qualified Network.HTTP.Types  as HTTP
+
+data HTTPConfig = HTTPConfig
+  { request :: HTTP.Request
+  , manager :: HTTP.Manager
+  }
+
+data LambdaResponse = LambdaResponse
+  { event     :: JSON.Value
+  , requestId :: RequestId
+  }
+  deriving stock Show
+
+newtype RequestId = RequestId Text
+  deriving stock (Show)
+
+instance Conversion BS.ByteString RequestId where
+  convert (RequestId requestId) = encodeUtf8 requestId
+
+data InternalLambdaClientError
+  = LambdaContextDecodeError Text
+  | MissingLambdaRunTimeApi
+  | InvalidLambdaRunTimeApi Text
+  | ConnectionError HTTP.HttpException
+  | ResponseBodyDecodeFailure Text
+  | UnSuccessfulEventRetrieval HTTP.Status Text
+  | UnSuccessfulEventSending HTTP.Status Text
+  | PayloadTooLarge Text
+  deriving stock Show
+
+instance Exception InternalLambdaClientError
+
+type LambdaClient = ExceptT InternalLambdaClientError IO
+
+getHttpConfig :: LambdaClient HTTPConfig
+getHttpConfig = do
+  awsLambdaRuntimeApi <- ExceptT $ maybeToEither MissingLambdaRunTimeApi
+    <$> lookupEnv "AWS_LAMBDA_RUNTIME_API"
+
+  req <- withExceptT (InvalidLambdaRunTimeApi . showc @Text)
+    . ExceptT
+    . try @_ @HTTP.HttpException
+    $ HTTP.parseRequest ("http://" <> awsLambdaRuntimeApi)
+
+  manager <- liftIO $ HTTP.newManager
+    $ HTTP.managerSetProxy HTTP.noProxy
+    $ HTTP.defaultManagerSettings
+        { HTTP.managerResponseTimeout     = HTTP.responseTimeoutNone
+        , HTTP.managerConnCount           = 1
+        , HTTP.managerIdleConnectionCount = 1
+        }
+  pure $ HTTPConfig req manager
+
+getNextLambdaResponse :: HTTPConfig -> LambdaClient LambdaResponse
+getNextLambdaResponse httpConfig = do
+  nextEvent   <- getNextEvent httpConfig
+  requestId <- RequestId <$> liftEither (getRequestId nextEvent)
+
+  pure LambdaResponse
+    { event = HTTP.responseBody nextEvent
+    , ..
+    }
+  where
+    getRequestId
+      :: HTTP.Response JSON.Value
+      -> Either InternalLambdaClientError Text
+    getRequestId response
+      = maybeToEither (LambdaContextDecodeError "Missing request Id")
+      $ getResponseHeader "Lambda-Runtime-Aws-Request-Id" response
+
+getNextEvent :: HTTPConfig -> LambdaClient (HTTP.Response JSON.Value)
+getNextEvent HTTPConfig{..} = do
+  response <- performHttpRequest $ HTTPConfig { request = toNextEventRequest request, .. }
+
+  let statusCode = HTTP.responseStatus response
+
+  unless (HTTP.statusIsSuccessful statusCode) . throwError
+    $ UnSuccessfulEventRetrieval statusCode "Could not retrieve next event"
+
+  pure response
+  where
+    toNextEventRequest :: HTTP.Request -> HTTP.Request
+    toNextEventRequest req = req
+      { HTTP.path = "2018-06-01/runtime/invocation/next"
+      }
+
+sendEvent
```

Comment:
I'd name this `sendEventResponse`, the path below suffixed `/response` makes this also a quite natural name.